src/arithmetic/algebraic_expression.c

```diff
@@ -524,20 +509,30 @@ AlgebraicExpression **AlgebraicExpression_FromQueryGraph(const QueryGraph *qg, u
 		return NULL;
 	}
 
+	bool acyclic = IsAcyclicGraph(qg);
```

Comment:
I hadn't thought of this before, but both our BFS and DFS routines are direction-agnostic, traversing incoming edges as well as outgoing. I would expect that `IsAcyclicGraph` needs to act similarly, no?

Reply:
No `IsAcyclicGraph` should conclude on the original user pattern.
`(a)-[]->(b)` doesn't contains a cycle, but if you like you can traverse from `b` to `a` on a reversed edge.